// medusa-interfaces/dist/base-service.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _medusaCoreUtils = require("medusa-core-utils");

var _typeorm = require("typeorm");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _instanceof(left, right) { if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) { return !!right[Symbol.hasInstance](left); } else { return left instanceof right; } }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _classCallCheck(instance, Constructor) { if (!_instanceof(instance, Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/**
 * Common functionality for Services
 * @interface
 */
var BaseService = /*#__PURE__*/function () {
  function BaseService() {
    _classCallCheck(this, BaseService);

    this.decorators_ = [];
  }

  _createClass(BaseService, [{
    key: "withTransaction",
    value: function withTransaction() {
      console.log("WARN: withTransaction called without custom implementation");
      return this;
    }
    /**
     * Used to build TypeORM queries.
     */

  }, {
    key: "buildQuery_",
    value: function buildQuery_(selector) {
      var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      var build = function build(obj) {
        var where = Object.entries(obj).reduce(function (acc, _ref) {
          var _ref2 = _slicedToArray(_ref, 2),
              key = _ref2[0],
              value = _ref2[1];

          // Undefined values indicate that they have no significance to the query.
          // If the query is looking for rows where a column is not set it should use null instead of undefined
          if (typeof value === "undefined") {
            return acc;
          }

          switch (true) {
            case _instanceof(value, _typeorm.FindOperator):
              acc[key] = value;
              break;

            case Array.isArray(value):
              acc[key] = (0, _typeorm.In)(_toConsumableArray(value));
              break;

            case value !== null && _typeof(value) === "object":
              var subquery = [];
              Object.entries(value).map(function (_ref3) {
                var _ref4 = _slicedToArray(_ref3, 2),
                    modifier = _ref4[0],
                    val = _ref4[1];

                switch (modifier) {
                  case "lt":
                    subquery.push({
                      operator: "<",
                      value: val
                    });
                    break;

                  case "gt":
                    subquery.push({
                      operator: ">",
                      value: val
                    });
                    break;

                  case "lte":
                    subquery.push({
                      operator: "<=",
                      value: val
                    });
                    break;

                  case "gte":
                    subquery.push({
                      operator: ">=",
                      value: val
                    });
                    break;

                  default:
                    acc[key] = value;
                    break;
                }
              });

              if (subquery.length) {
                acc[key] = (0, _typeorm.Raw)(function (a) {
                  return subquery.map(function (s, index) {
                    return "".concat(a, " ").concat(s.operator, " :").concat(index);
                  }).join(" AND ");
                }, subquery.map(function (s) {
                  return s.value;
                }));
              }

              break;

            default:
              acc[key] = value;
              break;
          }

          return acc;
        }, {});
        return where;
      };

      var query = {
        where: build(selector)
      };

      if ("deleted_at" in selector) {
        query.withDeleted = true;
      }

      if ("skip" in config) {
        query.skip = config.skip;
      }

      if ("take" in config) {
        query.take = config.take;
      }

      if ("relations" in config) {
        query.relations = config.relations;
      }

      if ("select" in config) {
        query.select = config.select;
      }

      if ("order" in config) {
        query.order = config.order;
      }

      return query;
    }
    /**
     * Confirms whether a given raw id is valid. Fails if the provided
     * id is null or undefined. The validate function takes an optional config
     * param, to support checking id prefix and length.
     * @param {string} rawId - the id to validate.
     * @param {object?} config - optional config
     * @returns {string} the rawId given that nothing failed
     */

  }, {
    key: "validateId_",
    value: function validateId_(rawId) {
      var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var prefix = config.prefix,
          length = config.length;

      if (!rawId) {
        throw new _medusaCoreUtils.MedusaError(_medusaCoreUtils.MedusaError.Types.INVALID_DATA, "Failed to validate id: ".concat(rawId));
      }

      if (prefix || length) {
        var _rawId$split = rawId.split("_"),
            _rawId$split2 = _slicedToArray(_rawId$split, 2),
            pre = _rawId$split2[0],
            rand = _rawId$split2[1];

        if (prefix && pre !== prefix) {
          throw new _medusaCoreUtils.MedusaError(_medusaCoreUtils.MedusaError.Types.INVALID_DATA, "The provided id: ".concat(rawId, " does not adhere to prefix constraint: ").concat(prefix));
        }

        if (length && length !== rand.length) {
          throw new _medusaCoreUtils.MedusaError(_medusaCoreUtils.MedusaError.Types.INVALID_DATA, "The provided id: ".concat(rawId, " does not adhere to length constraint: ").concat(length));
        }
      }

      return rawId;
    }
  }, {
    key: "shouldRetryTransaction",
    value: function shouldRetryTransaction(err) {
      var code = _typeof(err) === "object" ? String(err.code) : null;
      return code === "40001" || code === "40P01";
    }
    /**
     * Wraps some work within a transactional block. If the service already has
     * a transaction manager attached this will be reused, otherwise a new
     * transaction manager is created.
     * @param {function} work - the transactional work to be done
     * @param {string} isolation - the isolation level to be used for the work.
     * @return {any} the result of the transactional work
     */

  }, {
    key: "atomicPhase_",
    value: function () {
      var _atomicPhase_ = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(work, isolation) {
        var _this = this;

        var temp, doWork, result;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (!this.transactionManager_) {
                  _context2.next = 4;
                  break;
                }

                return _context2.abrupt("return", work(this.transactionManager_));

              case 4:
                temp = this.manager_;

                doWork = /*#__PURE__*/function () {
                  var _ref5 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(m) {
                    var result;
                    return regeneratorRuntime.wrap(function _callee$(_context) {
                      while (1) {
                        switch (_context.prev = _context.next) {
                          case 0:
                            _this.manager_ = m;
                            _this.transactionManager_ = m;
                            _context.prev = 2;
                            _context.next = 5;
                            return work(m);

                          case 5:
                            result = _context.sent;
                            _this.manager_ = temp;
                            _this.transactionManager_ = undefined;
                            return _context.abrupt("return", result);

                          case 11:
                            _context.prev = 11;
                            _context.t0 = _context["catch"](2);
                            _this.manager_ = temp;
                            _this.transactionManager_ = undefined;
                            throw _context.t0;

                          case 16:
                          case "end":
                            return _context.stop();
                        }
                      }
                    }, _callee, null, [[2, 11]]);
                  }));

                  return function doWork(_x3) {
                    return _ref5.apply(this, arguments);
                  };
                }();

                if (!isolation) {
                  _context2.next = 21;
                  break;
                }

                _context2.prev = 7;
                _context2.next = 10;
                return this.manager_.transaction(isolation, function (m) {
                  return doWork(m);
                });

              case 10:
                result = _context2.sent;
                return _context2.abrupt("return", result);

              case 14:
                _context2.prev = 14;
                _context2.t0 = _context2["catch"](7);

                if (!this.shouldRetryTransaction(_context2.t0)) {
                  _context2.next = 20;
                  break;
                }

                return _context2.abrupt("return", this.manager_.transaction(isolation, function (m) {
                  return doWork(m);
                }));

              case 20:
                throw _context2.t0;

              case 21:
                return _context2.abrupt("return", this.manager_.transaction(function (m) {
                  return doWork(m);
                }));

              case 22:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this, [[7, 14]]);
      }));

      function atomicPhase_(_x, _x2) {
        return _atomicPhase_.apply(this, arguments);
      }

      return atomicPhase_;
    }()
    /**
     * Dedicated method to set metadata.
     * @param {string} obj - the entity to apply metadata to.
     * @param {object} metadata - the metadata to set
     * @return {Promise} resolves to the updated result.
     */

  }, {
    key: "setMetadata_",
    value: function setMetadata_(obj, metadata) {
      var existing = obj.metadata || {};
      var newData = {};

      for (var _i2 = 0, _Object$entries = Object.entries(metadata); _i2 < _Object$entries.length; _i2++) {
        var _Object$entries$_i = _slicedToArray(_Object$entries[_i2], 2),
            key = _Object$entries$_i[0],
            value = _Object$entries$_i[1];

        if (typeof key !== "string") {
          throw new _medusaCoreUtils.MedusaError(_medusaCoreUtils.MedusaError.Types.INVALID_ARGUMENT, "Key type is invalid. Metadata keys must be strings");
        }

        newData[key] = value;
      }

      var updated = _objectSpread(_objectSpread({}, existing), newData);

      return updated;
    }
    /**
     * Adds a decorator to a service. The decorator must be a function and should
     * return a decorated object.
     * @param {function} fn - the decorator to add to the service
     */

  }, {
    key: "addDecorator",
    value: function addDecorator(fn) {
      if (typeof fn !== "function") {
        throw Error("Decorators must be of type function");
      }

      this.decorators_.push(fn);
    }
    /**
     * Runs the decorators registered on the service. The decorators are run in
     * the order they have been registered in. Failing decorators will be skipped
     * in order to ensure deliverability in spite of breaking code.
     * @param {object} obj - the object to decorate.
     * @return {object} the decorated object.
     */

  }, {
    key: "runDecorators_",
    value: function runDecorators_(obj) {
      var fields = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
      var expandFields = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
      return this.decorators_.reduce( /*#__PURE__*/function () {
        var _ref6 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(acc, next) {
          return regeneratorRuntime.wrap(function _callee3$(_context3) {
            while (1) {
              switch (_context3.prev = _context3.next) {
                case 0:
                  return _context3.abrupt("return", acc.then(function (res) {
                    return next(res, fields, expandFields);
                  })["catch"](function () {
                    return acc;
                  }));

                case 1:
                case "end":
                  return _context3.stop();
              }
            }
          }, _callee3);
        }));

        return function (_x4, _x5) {
          return _ref6.apply(this, arguments);
        };
      }(), Promise.resolve(obj));
    }
  }]);

  return BaseService;
}();

var _default = BaseService;
exports["default"] = _default;
